File: medbrains/apps/ios/Packages/MedBrainsOffline/Sources/MedBrainsOffline/OfflineAuthz.swift
import Foundation
import MedBrainsCore

/// Offline authorisation is a safety property and stays in Rust. This is the
/// thin Swift face of `AuthzCacheHandle` + `RevocationCacheHandle` +
/// `isActionOfflineRequired`, mirroring mobile-shell's `usePermissionCheck`:
/// an action the policy marks online-only is refused offline whatever the
/// cache says; otherwise the cache answers, then the JWT's permissions,
/// per the policy the caller names.
public final class OfflineAuthz: Sendable {
    public let authz: AuthzCacheHandle
    public let revocations: RevocationCacheHandle

    /// Same sizes the React Native shell opened: 1024 cache entries, 1 hour TTL, 4096 revocations.
    public init(directory: URL) throws {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        authz = try AuthzCacheHandle(path: directory.appendingPathComponent("authz.sled").path, capacity: 1024, defaultTtlSecs: 3600)
        revocations = try RevocationCacheHandle(path: directory.appendingPathComponent("revocations.sled").path, capacity: 4096)
    }

    public func check(_ key: CacheKey, jwtPermissions: [String], policy: OfflinePolicyKind = .cacheThenJwt) -> CheckOutcome {
        if isActionOfflineRequired(objectType: key.objectType, action: key.action) {
            return .deny(reason: .onlineRequired)
        }
        do {
            return try authz.checkOffline(key: key, jwtPermissions: jwtPermissions, policy: policy)
        } catch {
            // A fault is not a refusal, but offline it cannot be an allowance either.
            return .deny(reason: .cacheMissStrict)
        }
    }

    /// Record what the server answered, so the same question answers offline.
    public func remember(_ key: CacheKey, allowed: Bool) {
        try? authz.record(key: key, allowed: allowed, source: .cloudFresh)
    }
}

File: medbrains/apps/android/kit/src/main/kotlin/com/medbrains/kit/OfflineAuthz.kt
package com.medbrains.kit

import android.content.Context
import uniffi.edge_rn.AuthzCacheHandle
import uniffi.edge_rn.CacheKey
import uniffi.edge_rn.CacheSourceKind
import uniffi.edge_rn.CheckOutcome
import uniffi.edge_rn.DenyReasonKind
import uniffi.edge_rn.OfflinePolicyKind
import uniffi.edge_rn.RevocationCacheHandle
import uniffi.edge_rn.isActionOfflineRequired
import java.io.File

/**
 * Offline authorisation is a safety property and stays in Rust. This is the
 * thin Kotlin face of `AuthzCacheHandle` + `RevocationCacheHandle` +
 * `isActionOfflineRequired`, mirroring mobile-shell's `usePermissionCheck`:
 * an action the policy marks online-only is refused offline whatever the
 * cache says; otherwise the cache answers, then the JWT's permissions, per
 * the policy the caller names.
 */
class OfflineAuthz(context: Context) : AutoCloseable {
    private val dir = File(context.filesDir, "offline").apply { mkdirs() }

    /** Same sizes the React Native shell opened: 1024 cache entries, 1 hour TTL, 4096 revocations. */
    val authz = AuthzCacheHandle(File(dir, "authz.sled").path, 1024u, 3600uL)
    val revocations = RevocationCacheHandle(File(dir, "revocations.sled").path, 4096u)

    fun check(key: CacheKey, jwtPermissions: List<String>, policy: OfflinePolicyKind = OfflinePolicyKind.CACHE_THEN_JWT): CheckOutcome {
        if (isActionOfflineRequired(key.objectType, key.action)) {
            return CheckOutcome.Deny(DenyReasonKind.ONLINE_REQUIRED)
        }
        return runCatching { authz.checkOffline(key, jwtPermissions, policy) }
            // A fault is not a refusal, but offline it cannot be an allowance either.
            .getOrElse { CheckOutcome.Deny(DenyReasonKind.CACHE_MISS_STRICT) }
    }

    /** Record what the server answered, so the same question answers offline. */
    fun remember(key: CacheKey, allowed: Boolean) {
        runCatching { authz.record(key, allowed, CacheSourceKind.CLOUD_FRESH) }
    }

    override fun close() {
        authz.close()
        revocations.close()
    }
}
